### PR TITLE
Gh-388: Add "What is Python" section to user guide

### LINKED ISSUE #388: What is Python? - Additional section to User Guide
Following the restructuring of the documentation under https://github.com/gchq/gaffer-doc/issues/378 a page on Python to ensure that the end user / developer is aware of what Python is and how the project uses this technology.

Topics to include:

Basic overview of Python syntax (external doc links where applicable)
General usage of Python in the Gaffer API
Other config files that might use it

### PR DESCRIPTION
**Pull request by rj77259**

### CODE REVIEW

File: docs/user-guide/gaffer-basics/what-is-python.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# What is Python and how is it used in Gaffer?
+
+Python is a popular high level programming language that's seen massive popularity specifically in the data science and educational programming areas. Python is an interpreted general purpose language that's dynamically typed and garbage collected. For more info on Python please reference the official Python docs.
```
Comment (tb06904) [inline docs/user-guide/gaffer-basics/what-is-python.md]: Maybe add a link to the official python docs?

File: docs/user-guide/gaffer-basics/what-is-python.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+## Python in Gaffer
+
+Whilst Gaffer is written primarily in Java a Python interface has been provided so that you can programmatically access Gaffer functionality with Python, this can be accessed via the gafferpy library located in the gaffer-tools repository. This provides a Python 3.6+ compatible import that will allow you to speak directly to the Gaffer REST API, it supports persistent connections to Gaffer, connection via SSL and the associated python functionality to interact and extend Gaffer operations.
```
Comment (tb06904) [inline docs/user-guide/gaffer-basics/what-is-python.md]: Maybe a link to the gaffer-tools repo?

File: docs/user-guide/gaffer-basics/what-is-python.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    result = gc.execute_get(
+        g.GetSchema()
+    )
+
+    print('Schema:')
+    print(result)
+    print()
+```
+
+In this simple example gc passed in as a parameter is the gaffer_connector which orchestrates connection into Gaffer and g is the gafferpy's representation of the gaffer operations. This executes a get request into Gaffer to retrieve the current schema.
```
Comment (tb06904) [inline docs/user-guide/gaffer-basics/what-is-python.md]: Usually its handy to wrap extracts of code in back ticks so they standout in a text block e.g. like `gc` `gaffer_connector` and `g` here

File: docs/user-guide/gaffer-basics/what-is-python.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+A link to the gaffer tools repository can be found here: [GCHQ/Gaffer-Doc](https://github.com/gchq/gaffer-doc)
```
Comment (tb06904) [inline docs/user-guide/gaffer-basics/what-is-python.md]: This statement is incorrect its a link to the doc repo not tools repo

File: docs/user-guide/gaffer-basics/what-is-python.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+## Python in Gaffer
+
+Whilst Gaffer is written primarily in Java a Python interface has been provided so that you can programmatically access Gaffer functionality with Python, this can be accessed via the gafferpy library located in the gaffer-tools repository. This provides a Python 3.6+ compatible import that will allow you to speak directly to the Gaffer REST API, it supports persistent connections to Gaffer, connection via SSL and the associated python functionality to interact and extend Gaffer operations.
+
+Inside the gaffer-tools library you'll find a set of examples that show how you can interact with Gaffer, here is an example taken from that set:
+
+```py
+def get_schema(gc):
+    # Get Schema
+    result = gc.execute_get(
```
Comment (tb06904) [inline docs/user-guide/gaffer-basics/what-is-python.md]: The python PEP 8 only requires line wrap at 79 chars you could fit all this on one line

```python
result = gc.execute_get(g.GetSchema())
```

File: docs/user-guide/gaffer-basics/what-is-python.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+        g.GetSchema()
+    )
+
+    print('Schema:')
```
Comment (tb06904) [inline docs/user-guide/gaffer-basics/what-is-python.md]: You can use string formating here instead e.g.
```python
print("Schema:\n{0}\n".format(result))
```

**Review by tb06904 — APPROVED** [review]

File: docs/user-guide/gaffer-basics/what-is-python.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+        # Print result
+        print("Schema:\n{0}\n".format(result))
+
+    # Establish connection
+    g_connector = gaffer_connector.GafferConnector("http://localhost:8080/rest/latest")
+    get_schema(g_connector)
+    ```
+
+In this simple example, you can see the use of a `gaffer_connector` the purpose of this is to orchestrate the connection to a Gaffer REST endpoint. `g` is the main gaffer python module to allow access to various functions to run gaffer operations. This connection works by serialising the python code into JSON and then transferring this to be deserialised and ran in Gaffer.
```
Comment (GCHQDeveloper314) [inline docs/user-guide/gaffer-basics/what-is-python.md]:
```suggestion
In this simple example you can see the use of a `gaffer_connector`; the purpose of this is to orchestrate the connection to a Gaffer REST endpoint.  The main `gaffer` python module (usually imported as `g`) allows access to various functions to run Gaffer operations. This connection works by serialising the python code into JSON and then transferring this to be deserialised and ran in Gaffer.
```

**Review by GCHQDeveloper314 — COMMENTED** [review]
I'm wondering if all instances of "Python" should be capitalised. It's inconsistent at the moment. The [Python style guide](https://devguide.python.org/documentation/style-guide/#capitalization) says "The name of our favorite programming language is always capitalized."

Could also split some of the long lines.

**Comment by GCHQDeveloper314** [comment]
Would also be good to add a link to the GafferPy API page which is being added in #393. This can be done once that PR is merged.

**Review by GCHQDeveloper314 — APPROVED** [review]

**Pull request merged**